### PR TITLE
Drafted documentation for autograd. 

### PR DESCRIPTION
**Pull request by zackchase**
It's a little verbose and I can pare down but given the bare state of the page I think it's a big improvement over the status quo. 

I'd also like to add a simplest examples for with train_mode and with predict_mode. Probably the very most convincing would be just running an NDArrray through a dropout layer?

Pinging @piiswrong  and @szha

### CODE REVIEW

File: docs/api/python/autograd.md [hunk 1]
```diff
@@ -9,6 +9,78 @@
...
+>>> print(x.grad)
+>>>[ 2.  4.  6.  8.]
+<NDArray 4 @cpu(0)>
+```
+
+### ``Train_mode`` and ``predict_mode``
+
+Often, we want to define functions that behave differently
+when we are training models vs making predictions.
+By default, MXNet assumes we are in predict mode.
```
Comment (szha) [inline docs/api/python/autograd.md]: An example for dropout could be useful.

File: docs/api/python/autograd.md [hunk 2]
```diff
@@ -9,6 +9,78 @@
...
+However, usually when we take gradients, we are in the process of training.
+MXNet let's us decouple *training* vs *predict_mode* from
+*recording* vs *not recording*.
```
Comment (szha) [inline docs/api/python/autograd.md]: Users might not be aware of the coupling, since `record(train_mode=True)` is the default form of record.

File: docs/api/python/autograd.md [hunk 3]
```diff
@@ -9,6 +9,78 @@
...
+Additionally we can override the current score and elicit train or predict
+behaviors for some lines of code by placing them
+in a ``with autograd.train_mode():`` or ``with autograd.predict_mode():``
+block, respectively.
+This is especially for turning dropout layers on and off
+at prediction, as is popular for generating uncertainty estimates.
+
+
+Detailed tutorials are available in Part 1 of
+[the MXNet gluon book](http://gluon.mxnet.io/).
```
Comment (szha) [inline docs/api/python/autograd.md]: We should either add reference to the book in all gluon module description pages or none of them.

**Comment by zackchase** [comment]
Hi @szha, can you check the revised version?

**Comment by szha** [comment]
Looks good to me. @piiswrong ?

File: docs/api/python/autograd.md [hunk 4]
```diff
@@ -9,6 +9,64 @@
...
+### Record vs Pause
+
+`autograd` records computation history on the fly to calculate gradients later.
+This is only enabled inside a `with autograd.record():` block.
+A `with auto_grad.pause()` block can be used inside a `record()` block
+to temporarily disable recording.
+
+To compute gradient with respect to an `NDArray` `x`, first call `x.attach_grad()`
+to allocate space for the gradient. Then, start a `with autograd.record()` block,
+and do some computation. Finally, call `backward()` on the result:
```
Comment (fhieber) [inline docs/api/python/autograd.md]: The following example does not seem to call a method called "backward()".

File: docs/api/python/autograd.md [hunk 5]
```diff
@@ -9,6 +9,64 @@
...
+[ 2.  4.  6.  8.]
+<NDArray 4 @cpu(0)>
+```
+
+
+## Train mode and Predict Mode
+
+Some operators (Dropout, BatchNorm, etc) behave differently in
+when training and when making predictions.
+This can be controled with `train_mode` and `predict_mode` scope.
```
Comment (fhieber) [inline docs/api/python/autograd.md]: typo: controled -> controlled

File: docs/api/python/autograd.md [hunk 6]
```diff
@@ -9,6 +9,64 @@
...
+...     y = x * x + 1
+>>> print(x.grad)
+[ 2.  4.  6.  8.]
+<NDArray 4 @cpu(0)>
+```
+
+
+## Train mode and Predict Mode
+
+Some operators (Dropout, BatchNorm, etc) behave differently in
```
Comment (fhieber) [inline docs/api/python/autograd.md]: -in

**Comment by domdivakaruni** [comment]
this is awesome! Thanks guys
